index.js

```diff
@@ -600,6 +628,94 @@ function readOpsBulk(stream, callback) {
   });
 }
 
+ShareDbMongo.prototype._getOpLink = function(collectionName, id, to, callback) {
+  if (!this.fasterGetOps) return this._getSnapshotOpLink(collectionName, id, callback);
+
+  var db = this;
+  this.getOpCollection(collectionName, function (error, collection) {
+    if (error) return callback(error);
+
+    var query = {
+      d: id,
+      v: { $gte: to }
+    };
+
+    var projection = {
+      _id: 0,
+      v: 1,
+      o: 1
+    };
+
+    var cursor = collection.find(query).sort({ v: 1 }).project(projection);
+
+    getFirstOpWithUniqueVersion(cursor, function (error, op) {
+      if (error) return callback(error);
+      if (op) return callback(null, { _o: op.o, _v: op.v });
+
+      // If we couldn't find an op to link back from, then fall back to using the current
+      // snapshot, which is guaranteed to have a link to a valid op.
+      db._getSnapshotOpLink(collectionName, id, callback);
+    });
+  });
+};
+
+// When getting ops, we need to consider the case where an op is committed to the database,
+// but its application to the snapshot is subsequently rejected. This can leave multiple ops
+// with the same values for 'd' and 'v', and means that we may return multiple ops for a single
+// version if we just perform a naive 'find' operation.
+// To avoid this, we try to fetch the first op after 'to' which has a unique 'v', and then we
+// work backwards from that op using the linked op 'o' field to get a valid chain of ops.
+function getFirstOpWithUniqueVersion(cursor, ops, callback) {
+  if (typeof ops === 'function') {
+    callback = ops;
+    ops = [];
+  }
+
+  // We have to look back at the last 3 ops to determine if the version in the middle was
+  // unique. For example, if we get 1 -> 2 -> 3 then we know that v2 is unique (and we can't
+  // make the same assumption about v3, because the next op we fetch might also be v3).
+  var currentOp = ops[ops.length - 1]
+  var previousOp = ops[ops.length - 2];
+  var oneBeforePreviousOp = ops[ops.length - 3];
+
+  var currentVersion = currentOp && currentOp.v;
+  var previousVersion = previousOp && previousOp.v;
+  var oneBeforePreviousVersion = oneBeforePreviousOp && oneBeforePreviousOp.v;
+
+  var previousVersionWasUnique = typeof previousVersion === 'number'
+    && previousVersion !== currentVersion
+    && previousVersion !== oneBeforePreviousVersion;
+
+  if (previousVersionWasUnique) {
```

Comment:
Turn this check into a function to shorten this function.

Reply:
Alternatively, we should pull this out into its own class, who has only these three objects (so that we don't keep the rest of the array of ops in memory that we don't need).